cxx-qt-gen/src/gen_rs.rs

```diff
@@ -669,6 +730,79 @@ fn generate_property_methods_rs(obj: &QObject) -> Result<Vec<TokenStream>, Token
     Ok(property_methods)
 }
 
+fn generate_signal_methods_rs(obj: &QObject) -> Result<Vec<TokenStream>, TokenStream> {
+    let mut signal_methods = Vec::new();
+    let mut queued_cases = Vec::new();
+    let mut immediate_cases = Vec::new();
+
+    for signal in &obj.signals {
+        let emit_ident = &signal.emit_ident.rust_ident;
+        let enum_ident = &signal.enum_ident;
+        let parameters = signal
+            .parameters
+            .iter()
+            .map(|parameter| &parameter.ident)
+            .collect::<Vec<&Ident>>();
+        let parameters_to_value_immediate = signal
+            .parameters
+            .iter()
+            .map(|parameter| {
+                let ident = &parameter.ident;
+                if parameter.type_ident.qt_type.is_opaque() {
+                    quote! { &#ident.to_unique_ptr() }
+                } else if parameter.type_ident.qt_type.is_ref() {
+                    quote! { &#ident }
+                } else {
+                    ident.into_token_stream()
+                }
+            })
+            .collect::<Vec<TokenStream>>();
+        let parameters_to_value_queued = signal
+            .parameters
+            .iter()
+            .map(|parameter| {
+                let ident = &parameter.ident;
+                if parameter.type_ident.qt_type.is_opaque() {
+                    quote! { #ident.to_unique_ptr() }
+                } else {
+                    ident.into_token_stream()
+                }
+            })
+            .collect::<Vec<TokenStream>>();
+        let signal_ident = &signal.signal_ident.rust_ident;
+
+        queued_cases.push(quote! {
+            Signal::#enum_ident { #(#parameters),* } => self.cpp.as_mut().#emit_ident(#(#parameters_to_value_queued),*),
+        });
+
+        immediate_cases.push(quote! {
+            Signal::#enum_ident { #(#parameters),* } => self.cpp.as_mut().#signal_ident(#(#parameters_to_value_immediate),*),
+        });
+    }
+
+    if !queued_cases.is_empty() {
+        signal_methods.push(quote! {
+            pub fn emit_queued(&mut self, signal: Signal) {
+                match signal {
+                    #(#queued_cases)*
+                }
+            }
+        });
+    }
+
+    if !immediate_cases.is_empty() {
+        signal_methods.push(quote! {
+            pub unsafe fn emit_immediate(&mut self, signal: Signal) {
```

Comment:
If we use a `std::recursive_mutex` instead of a normal mutex to protect the RustObj, then can `unsafe` be removed here ? As then if a signal causes an invokable to be called from the same thread that is "fine" ? Or does this conflict with Rust's ownership/threading rules as it can't tell that halfway through invokable A we will jump to invokable B then back to A ?

https://en.cppreference.com/w/cpp/thread/recursive_mutex